src/libcore/char.rs

```diff
@@ -654,3 +654,38 @@ impl Iterator for EncodeUtf16 {
         self.as_slice().iter().size_hint()
     }
 }
+
+
+/// An iterator over an iterator of bytes of the characters the bytes represent
+/// as UTF-8. Invalid sequences become `None`.
+#[unstable(feature = "decode_utf8", issue = "33906")]
+#[derive(Clone, Debug)]
+pub struct DecodeUtf8<I: Iterator<Item = u8>>(I);
+
+#[unstable(feature = "decode_utf8", issue = "33906")]
+impl<I: Iterator<Item = u8>> From<I> for DecodeUtf8<I> {
+    #[inline]
+    fn from(i: I) -> Self { DecodeUtf8(i) }
+}
+
+#[unstable(feature = "decode_utf8", issue = "33906")]
+impl<I: Iterator<Item = u8>> Iterator for DecodeUtf8<I> {
+    type Item = Option<char>;
```

Comment:
Idiomatically we tend to prefer to return a `Result` rather than an `Option` to indicate an error. This helps us give an entry point for an opaque struct to attach error information to.